File: LeetCodeRecord/MethodFirst/202-IsHappy.java
package MethodFirst;

import java.util.HashSet;

class IsHappy {
    public int getNext(int oldNum) {
        int newNum = 0;
        while (oldNum != 0) {
            int temp = oldNum % 10;
            newNum += temp * temp;
            oldNum = oldNum / 10;
        }

        return newNum;
    }

    public boolean isHappy(int n) {
        int slow = n;
        int fast = n;
        
        do {
            slow = getNext(slow);
            fast = getNext(fast);
            fast = getNext(fast);
        } while(slow != fast);

        return slow == 1;
    }

    public boolean isHappy2(int n) {
        HashSet<Integer> nums = new HashSet<Integer>();

        while (n != 1 && !nums.contains(n)) {
            nums.add(n);            
            n = getNext(n);
        }

        return n == 1;
    }
}

File: LeetCodeRecord/ACFirst-Golang/202-IsHappy.go
package main

func isHappy(n int) bool {
	slow := n
	fast := n

	for fast != 1 {
		slow = doSum(slow)
		fast = doSum(fast)
		fast = doSum(fast)

		if slow == fast {
			break
		}
	}

	if fast == 1 {
		return true
	}

	return false
}

func doSum(n int) int {
	sum := 0
	for n != 0 {
		temp := n % 10
		sum += temp * temp
		n = n / 10
	}

	return sum
}

func isHappy2(n int) bool {
	nums := make(map[int]bool)
	nums[n] = true

	for n != 1 {
		n = doSum(n)
		if _, ok := nums[n]; ok {
			break
		}

		nums[n] = true
	}

	if n == 1 {
		return true
	}

	return false
}
